src/OpenTelemetry.Contrib.Instrumentation.AWS/Implementation/SnsRequestContextAdapter.cs

```diff
@@ -0,0 +1,89 @@
+// <copyright file="SnsRequestContextAdapter.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.Runtime;
+using Amazon.Runtime.Internal;
+using Amazon.SimpleNotificationService.Model;
+
+namespace OpenTelemetry.Contrib.Instrumentation.AWS.Implementation;
+internal class SnsRequestContextAdapter : IRequestContextAdapter
+{
+    // SQS/SNS message attributes collection size limit according to
+    // https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html and
+    // https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+    private const int MaxMessageAttributes = 10;
+
+    private readonly ParameterCollection? parameters;
+    private readonly PublishRequest? originalRequest;
+
+    public SnsRequestContextAdapter(IRequestContext context)
+    {
+        this.parameters = context.Request?.ParameterCollection;
+        this.originalRequest = context.OriginalRequest as PublishRequest;
+    }
+
+    public bool CanInject => this.originalRequest?.MessageAttributes != null && this.parameters != null;
+
+    public void AddAttributes(IReadOnlyDictionary<string, string> attributes)
+    {
+        if (!this.CanInject)
```

Comment:
This check feels redundant. If we are very worried about users forgetting to call CanInject, we could remove `CanInject` and instead create a factory method that returns null instead of a new instance if CanInject would be false.

At that point (maybe even now already), it wouldn't really be an adapter anymore but basically just a single method that could as well be static.